### ISSUE TITLE
404 errors returned by Envoy proxies and Virtual Gateway

### ISSUE THREAD

**Issue by benmurden** [aws/aws-app-mesh-roadmap#447]
**Summary**
We have suddenly started seeing Envoy return 404 errors for services and paths that used to work, without any apparent recent config changes. This is happening on our service proxies and on the Virtual Gateway.

**Steps to Reproduce**
Create an App Mesh and run it for a few months.

**Are you currently working around this issue?**
It seems if we restart the ECS tasks in the chain that fails we can restore functionality for a while, but this has recently started happening more frequently after we had been running for several months without issue.

**Additional context**
- App Mesh cluster uses ECS Fargate for Virtual Gateway and services.
- Both the Virtual Gateway with external requests and Envoy proxies serving internal requests have started doing this.
- So far this only seems to be happening on one staging environment out of four accounts where we use App Mesh.

**Attachments**
Logs from one such proxy.
```
[2022-11-14T03:07:48.082Z] "GET /health/ready HTTP/1.1" 404 NR 0 0 0 - "-" "Envoy/HC" "1005b212-59a7-97ad-8ba8-7d1672f8e350" "internal-service.app.local:8080" "-"
```

We tried testing if the application was actually still up by binding it to a port that was not covered by Envoy and using Curl:
```sh
# Test on the port proxied by Envoy
root@ip-10-2-86-178:/app# curl -v internal-service.app.local:8080/health/ready
*   Trying 10.2.97.133:8080...
* Connected to internal-service.app.local (10.2.97.133) port 8080 (#0)
> GET /health/ready HTTP/1.1
> Host: internal-service.app.local:8080
> User-Agent: curl/7.74.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< date: Mon, 14 Nov 2022 04:33:36 GMT
< server: envoy
< content-length: 0
< x-envoy-upstream-service-time: 2
<
* Connection #0 to host internal-service.app.local left intact
# Test on the port not proxied by Envoy (direct connection)
root@ip-10-2-86-178:/app# curl -v internal-service.app.local:8081/health/ready
*   Trying 10.2.97.133:8081...
* Connected to internal-service.app.local (10.2.97.133) port 8081 (#0)
> GET /health/ready HTTP/1.1
> Host: internal-service.app.local:8081
> User-Agent: curl/7.74.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=utf-8
< Date: Mon, 14 Nov 2022 04:34:02 GMT
< Content-Length: 16
<
{"status":"ok"}
* Connection #0 to host internal-service.app.local left intact
```

**Comment by suniltheta** [comment]
Would you be able to raise a support ticket on this issue ? If not fell free to ping me on App Mesh slack https://github.com/aws/aws-app-mesh-roadmap#slack-community suniltheta@. I would like to collect more details on the issue that you are facing.

**Comment by burakvp** [comment]
Not sure if this is related but I had similar issue with app mesh gateway when requests was going from inside mesh via nginx proxy_pass. 
The issue was caused by "x-app-mesh-virtual-service" header preserved by proxy_pass and additional one attached by sidecar for egress traffic. So
`curl https://<gw_hostname>/ -H "x-app-mesh-virtual-service: proxy.ecs-demo.us-east-1.local" was failing. but
`curl https://<gw_hostname>/ ` was passing. It seems like somehow `x-app-mesh-virtual-service` breaks proper routing.

**Comment by suniltheta** [comment]
@burakvp  I have not come across this issue that you just mentioned. Also I am not very clear about the header that is used in the curl call. If you are still running into the problem, please feel free to open a support ticket with more details of your issue and how to recreate it.

---

Meanwhile after discussing with benmurden@, this issue happens intermittently while using EFS mount. I will go ahead and close this issue as we are not taking any immediate action item on this. If anyone else is running into similar problem, please create a support ticket on AWS App Mesh.

**Comment by benmurden** [comment]
Thanks @suniltheta, I'll just add a little more context for anyone else who happens to find this issue.

Based on one of the [AWS App Mesh examples](https://github.com/aws/aws-app-mesh-examples/tree/main/walkthroughs/howto-mutual-tls-file-provided#step-3-export-our-custom-envoy-image), we put together a custom Envoy image which seems to have worked fine, but the problems started when we followed this suggestion in the Readme:
> Note: This walkthrough uses this custom Envoy image for illustration purposes; using this method, rotating certificates will require a deployment of your tasks. You may instead choose to mount to your tasks an Elastic Block Storage or Elastic File System volume to store your TLS materials and reference them through the API accordingly.

We implemented an EFS mount volume on our Envoy sidecar containers in ECS in order for any certificate changes to propagate to other Envoy instances. This seems to have been the point at which we started seeing these intermittent problems, and since removing the EFS volume mount, we have not seen the same issue.
